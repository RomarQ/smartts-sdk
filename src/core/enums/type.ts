export enum TypeAtom {
    record = 'record',
    variant = 'variant',
    chain_id = 'chain_id',
    unit = 'unit',
    nat = 'nat',
    int = 'int',
    mutez = 'mutez',
    bool = 'bool',
    bytes = 'bytes',
    string = 'string',
    address = 'address',
    timestamp = 'timestamp',
    bls12_381_fr = 'bls12_381_fr',
    bls12_381_g1 = 'bls12_381_g1',
    bls12_381_g2 = 'bls12_381_g2',
    key = 'key',
    key_hash = 'key_hash',
    list = 'list',
    set = 'set',
    option = 'option',
    map = 'map',
    big_map = 'bigmap',
    signature = 'signature',
    operation = 'operation',
    never = 'never',
    tuple = 'tuple',
    lambda = 'lambda',
    ticket = 'ticket',
    sapling_state = 'sapling_state',
    contract = 'contract',
    sapling_transaction = 'sapling_transaction',
}

export default TypeAtom;
