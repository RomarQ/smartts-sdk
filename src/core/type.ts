import { IToString } from '../typings';

export class BaseType {
    private _ = 'Type';
}

export class SimpleType extends BaseType {
    constructor(public name: string) {
        super();
    }

    toString() {
        return `"${this.name}"`;
    }
}

export class T_TList extends BaseType {
    constructor(public innerType: IToString) {
        super();
    }

    toString() {
        return `(list ${this.innerType})`;
    }
}

export const TList = (innerType: BaseType) => new T_TList(innerType);

export const TNat = new SimpleType('nat');
export const TInt = new SimpleType('int');
export const TMutez = new SimpleType('mutez');
export const TString = new SimpleType('string');
