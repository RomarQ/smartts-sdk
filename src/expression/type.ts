import type { IExpression } from '../typings/expression';
import type { IType } from '../typings/type';
import { Expression } from '../core/expression';
import ExpressionAtom from '../core/enums/expression';
import { LineInfo } from '../misc/utils';
import { proxy } from '../misc/proxy';

/**
 * Annotate an expression with type.
 *
 * ```typescript
 * AsType(GetVariable("some_variable"), TNat());
 * ```
 *
 * @param expression
 * @param type
 * @param {LineInfo} line Source code line information (Used in error messages)
 *
 * @category | Type Handling
 *
 * @returns {IExpression} An expression
 */
export const AsType = (expression: IExpression, type: IType, line = new LineInfo()): IExpression<any> =>
    proxy(new Expression(ExpressionAtom.type_annotation, expression, type, line), Expression.proxyHandler);
