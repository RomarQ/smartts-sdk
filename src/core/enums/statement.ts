export enum StatementAtom {
    verify = 'verify',
    ifBlock = 'ifBlock',
    elseBlock = 'elseBlock',
    forGroup = 'forGroup',
    whileBlock = 'whileBlock',
    set = 'set',
    defineLocal = 'defineLocal',
    result = 'result',
    failwith = 'failwith',
    match = 'match',
    match_cases = 'match_cases',
    set_type = 'set_type',
}

export default StatementAtom;
