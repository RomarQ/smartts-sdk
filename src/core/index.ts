import type { IToString, IToType } from '../typings';
import type { ILiteral } from '../typings/literal';
import type { IType } from '../typings/type';

import { Expression } from './expression';
import { Mutez } from './literal';
import Utils, { LineInfo } from '../misc/utils';
import { TUnit } from './type';

export class Flag {
    args: string[] = [];

    constructor(public flag: string, ...args: string[]) {
        this.args = args;
    }

    toString() {
        return `(${this.flag}${(' ' + this.args.join(' ')).trim()})`;
    }
}

interface EntryPointOptions {
    mock?: boolean;
    lazy?: boolean;
    lazyAndCodeless?: boolean;
}
export class EntryPoint {
    options = {
        mock: false,
        lazy: false,
        lazyAndCodeless: false,
    };
    inType?: IType;
    statements: IToString[] = [];

    constructor(public name: string, public line = new LineInfo()) {}

    config(options?: EntryPointOptions) {
        if (options?.mock) {
            this.options.mock = options.mock;
        }
        if (options?.lazy) {
            this.options.lazy = options.lazy;
        }
        if (options?.lazyAndCodeless) {
            this.options.lazyAndCodeless = options.lazyAndCodeless;
        }
        return this;
    }

    inputType(type: IType) {
        this.inType = type;
        return this;
    }

    code(callback: (arg: IToString) => IToString[]) {
        this.statements = callback(new Expression('params', new LineInfo()));
        return this;
    }

    toString() {
        const notMock = Utils.capitalizeBoolean(!this.options.mock);
        const isLazy = Utils.capitalizeBoolean(this.options.lazy);
        const isLazyAndCodeless = Utils.capitalizeBoolean(this.options.lazyAndCodeless);
        const hasParams = Utils.capitalizeBoolean(!!this.inType);
        return `(${this.name} ${notMock} ${isLazy} ${isLazyAndCodeless} ${hasParams} ${
            this.line
        } (${this.statements.join(' ')}))`;
    }
}

export class Contract {
    options: {
        initialBalance: ILiteral;
        flags: Flag[];
    } = {
        initialBalance: Mutez(0),
        flags: [],
    };
    public initialStorage: IToString & IToType;
    public entries: EntryPoint[];

    constructor(
        args: {
            initialStorage: IToString & IToType;
            entries: EntryPoint[];
        },
        public line = new LineInfo(),
    ) {
        this.initialStorage = args.initialStorage;
        this.entries = args.entries;
    }

    config(options?: { initialBalance?: ILiteral; flags?: Flag[] }) {
        if (options?.flags) {
            this.options.flags = options.flags;
        }
        if (options?.initialBalance) {
            this.options.initialBalance = options.initialBalance;
        }
        return this;
    }

    toString() {
        return `
        (
            template_id (static_id 0 ${this.line})
            storage ${this.initialStorage.toString()}
            storage_type (${this.initialStorage.toType()})
            messages (${this.entries.map((entry) => entry.toString()).join(' ')})
            flags (${this.options.flags.map((flag) => flag.toString())})
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance ${this.options.initialBalance}
        )
        `;
    }
}

export { default as Command } from './command';
export { default as Expression } from './expression';
export { default as Type } from './type';
export { default as Literal } from './literal';
export * from './blockchain_operations';
