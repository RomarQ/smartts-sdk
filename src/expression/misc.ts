import ExpressionAtom from '../core/enums/expression';
import { Expression } from '../core/expression';
import { proxy } from '../misc/proxy';
import { LineInfo } from '../misc/utils';
import { IExpression } from '../typings/expression';
import { List } from './literal';
import ValueAtom from '../core/enums/literal';

/**
 * Concatenate a list with values of type `TString()` or `TBytes()`.
 *
 * ```typescript
 * Concat([ String("Hello"), String(" "), String("World") ]);
 * ```
 *
 * @category | Concatenation
 *
 * @param values
 * @param {LineInfo} line Source code line information (Used in error messages)
 *
 * @returns {IExpression} An expression.
 */
export const Concat = <T extends ValueAtom.bytes | ValueAtom.string>(values: IExpression<T>[], line = new LineInfo()) =>
    proxy(new Expression<T>(ExpressionAtom.concat, List(values), line), Expression.proxyHandler);

/**
 * Obtain size of values with type `TString()`, `TBytes()`, `TList(...)`, `TSet(...)` and `TMap()`.
 *
 * ```typescript
 * SizeOf(String("Hello")); // Nat(5)
 * ```
 *
 * @category | Size
 *
 * @param value
 * @param {LineInfo} line Source code line information (Used in error messages)
 *
 * @returns {IExpression} An expression.
 */
export const SizeOf = (
    value: IExpression<ValueAtom.bytes | ValueAtom.string | ValueAtom.list | ValueAtom.set | ValueAtom.map>,
    line = new LineInfo(),
) => proxy(new Expression<ValueAtom.nat>(ExpressionAtom.size, value, line), Expression.proxyHandler);
