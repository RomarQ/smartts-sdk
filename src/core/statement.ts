import type { IExpression } from '../typings/expression';
import type { ILiteral } from '../typings/literal';
import type { IStatement } from '../typings/statement';
import type { IType } from '../typings/type';
import { LineInfo } from '../misc/utils';
import StatementAtom from './enums/statement';
import { Expression } from './expression';
import ExpressionAtom from './enums/expression';
import { Proxied, proxy } from '../misc/proxy';
import { Add, GetVariable, Iterator, LessThanOrEqual } from '../expression';
import { NewVariable, SetValue } from '../statement';

export class Statement implements IStatement {
    args;

    constructor(public name: string, ...args: (IExpression | IType | LineInfo | string | ILiteral)[]) {
        this.args = args || [];
    }

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}

export class IfStatment implements IStatement {
    #condition: IExpression;
    #thenStatements: IStatement[];
    #elseStatements: IStatement[];
    #line: LineInfo;

    constructor(
        condition: IExpression,
        line = new LineInfo(),
        thenStatements?: IStatement[],
        elseStatements?: IStatement[],
    ) {
        this.#condition = condition;
        this.#thenStatements = thenStatements || [];
        this.#elseStatements = elseStatements || [];
        this.#line = line;
    }

    public Then(statements: IStatement[]) {
        this.#thenStatements = statements;
        return this;
    }

    public Else(statements: IStatement[]) {
        this.#elseStatements = statements;
        return this;
    }

    private ifBlock() {
        return `(${StatementAtom.ifBlock} ${this.#condition} (${this.#thenStatements.join(' ')}) ${this.#line})`;
    }

    private elseBlock() {
        return `(${StatementAtom.elseBlock} (${this.#elseStatements.join(' ')}))`;
    }

    [Symbol.toPrimitive]() {
        let expression = this.ifBlock();
        if (this.#elseStatements.length) {
            expression = `${expression} ${this.elseBlock()}`;
        }
        return expression;
    }
}

export class VariantMatchStatement implements IStatement {
    static idCounter = 0;
    private cases: Record<string, IStatement[]> = {};

    static get nextID() {
        return ++VariantMatchStatement.idCounter;
    }

    constructor(private variant: IExpression, private argumentName: string, private line = new LineInfo()) {}

    public setArgumentName(argumentName: string): this {
        this.argumentName = argumentName;
        return this;
    }

    public Case(branch: string, buildStatements: (arg: Proxied<IExpression>) => IStatement[]): this {
        const variantArgument = this.variantArgument(`${this.argumentName}_${branch}`, this.line);
        this.cases[branch] = buildStatements(variantArgument);
        return this;
    }

    private caseArgument(argumentName: string, line: LineInfo) {
        return new Expression(ExpressionAtom.cases_arg, `"${argumentName}"`, line);
    }

    private variantArgument(argumentName: string, line: LineInfo) {
        return proxy(new Expression(ExpressionAtom.variant_arg, `"${argumentName}"`, line), Expression.proxyHandler);
    }

    [Symbol.toPrimitive]() {
        const caseArgument = this.caseArgument(this.argumentName, this.line);
        const matchCases = Object.keys(this.cases).map((branch) => {
            return new Statement(
                StatementAtom.match,
                caseArgument,
                branch,
                `"${this.argumentName}_${branch}"`,
                `(${this.cases[branch].join(' ')})`,
                this.line,
            );
        });
        return `(${StatementAtom.match_cases} ${this.variant} "${this.argumentName}" (${matchCases.join(' ')}) ${
            this.line
        })`;
    }
}

export class WhileStatement implements IStatement {
    constructor(private condition: IExpression, private statements: IStatement[] = [], private line = new LineInfo()) {}

    public Do(statements: IStatement[]) {
        this.statements = statements;
        return this;
    }

    [Symbol.toPrimitive]() {
        return `(${StatementAtom.whileBlock} ${this.condition} (${this.statements.join(' ')}) ${this.line})`;
    }
}

export class ForStatement implements IStatement {
    static idCounter = 0;

    static get nextID() {
        return ++VariantMatchStatement.idCounter;
    }

    constructor(
        private from: IExpression,
        private to: IExpression,
        private increment: IExpression,
        private statements: IStatement[] = [],
        private iteratorName: string,
        private line = new LineInfo(),
    ) {}

    public setIteratorName(iteratorName: string): this {
        this.iteratorName = iteratorName;
        return this;
    }

    public Do(callback: (iterator: Proxied<IExpression>) => IStatement[], line = new LineInfo()) {
        const iterator = GetVariable(this.iteratorName, line);
        this.statements = callback(iterator);
        return this;
    }

    [Symbol.toPrimitive]() {
        const variable = NewVariable(this.iteratorName, this.from);
        const condition = LessThanOrEqual(GetVariable(this.iteratorName), this.to);
        const stmts = [
            ...this.statements,
            SetValue(GetVariable(this.iteratorName), Add(GetVariable(this.iteratorName), this.increment)),
        ];
        return `${variable} (${StatementAtom.whileBlock} ${condition} (${stmts.join(' ')}) ${this.line})`;
    }
}

export class ForEachStatement implements IStatement {
    static idCounter = 0;

    static get nextID() {
        return ++VariantMatchStatement.idCounter;
    }

    constructor(
        private list: IExpression,
        private statements: IStatement[],
        private iteratorName: string,
        private line = new LineInfo(),
    ) {}

    public setIteratorName(iteratorName: string): this {
        this.iteratorName = iteratorName;
        return this;
    }

    public Do(callback: (iterator: Proxied<IExpression>) => IStatement[], line = new LineInfo()) {
        const iterator = Iterator(this.iteratorName, line);
        this.statements = callback(iterator);
        return this;
    }

    [Symbol.toPrimitive]() {
        return `(${StatementAtom.forGroup} "${this.iteratorName}" ${this.list} (${this.statements.join(' ')}) ${
            this.line
        })`;
    }
}
