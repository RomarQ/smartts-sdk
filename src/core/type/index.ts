import { composeRightCombLayout, parenthesis } from '../../misc/utils';
import { ILayout } from '../../typings/literal';
import { IType } from '../../typings/type';
import { Layout } from '../enums/layout';
import TypeAtom from '../enums/type';

export class SimpleType implements IType {
    _isType = true as const;

    constructor(public name: string) {}

    toString() {
        return `"${this.name}"`;
    }
}

export class ContainerType implements IType {
    _isType = true as const;

    constructor(public type: string, public innerTypes: (IType | number)[]) {}

    toString() {
        return `(${this.type} ${this.innerTypes.join(' ')})`;
    }
}

export class Type_Record implements IType {
    _isType = true as const;

    private layout: ILayout | Layout;

    constructor(private atom: string, private fields: Record<string, IType>, layout?: ILayout | Layout) {
        this.layout = layout || composeRightCombLayout(Object.keys(fields));
    }

    private translateLayout = (layout: ILayout | Layout): string => {
        const normalizeTuple = (layout: string | ILayout): string => {
            if (typeof layout === 'string') {
                return `("${layout}")`;
            }
            const [left, right] = layout;
            return parenthesis([normalizeTuple(left), normalizeTuple(right)].join(' '));
        };

        if (Array.isArray(layout)) {
            return layout.length ? `(Some ${normalizeTuple(layout)})` : 'None';
        }
        return '(Some Right)';
    };

    private buildFields = (fields: Record<string, IType>) => {
        return Object.entries(fields).map(([field, ty]) => `(${field} ${ty.toString()})`);
    };

    toString() {
        return `(${this.atom} (${this.buildFields(this.fields).join(' ')}) ${this.translateLayout(this.layout)})`;
    }
}

export const TUnknown: IType = {
    _isType: true as const,
    toString: () => '(unknown 0)',
};
// Singleton types
export const TUnit = () => new SimpleType(TypeAtom.unit);
export const TNat = () => new SimpleType(TypeAtom.nat);
export const TInt = () => new SimpleType(TypeAtom.int);
export const TMutez = () => new SimpleType(TypeAtom.mutez);
export const TString = () => new SimpleType(TypeAtom.string);
export const TBool = () => new SimpleType(TypeAtom.bool);
export const TAddress = () => new SimpleType(TypeAtom.address);
export const TTimestamp = () => new SimpleType(TypeAtom.timestamp);
export const TChain_id = () => new SimpleType(TypeAtom.chain_id);
export const TBytes = () => new SimpleType(TypeAtom.bytes);
export const TBls12_381_fr = () => new SimpleType(TypeAtom.bls12_381_fr);
export const TBls12_381_g1 = () => new SimpleType(TypeAtom.bls12_381_g1);
export const TBls12_381_g2 = () => new SimpleType(TypeAtom.bls12_381_g2);
export const TKey = () => new SimpleType(TypeAtom.key);
export const TKey_hash = () => new SimpleType(TypeAtom.key_hash);
export const TSignature = () => new SimpleType(TypeAtom.signature);
export const TOperation = () => new SimpleType(TypeAtom.operation);
export const TNever = () => new SimpleType(TypeAtom.never);
// Container types
export const TList = (innerType: IType) => new ContainerType(TypeAtom.list, [innerType]);
export const TSet = (innerType: IType) => new ContainerType(TypeAtom.set, [innerType]);
export const TOption = (innerType: IType) => new ContainerType(TypeAtom.option, [innerType]);
export const TMap = (keyType: IType, valueType: IType) => new ContainerType(TypeAtom.map, [keyType, valueType]);
export const TBig_map = (keyType: IType, valueType: IType) => new ContainerType(TypeAtom.big_map, [keyType, valueType]);
export const TPair = (left: IType, right: IType) => new ContainerType(TypeAtom.tuple, [left, right]);
export const TLambda = (left: IType, right: IType) => new ContainerType(TypeAtom.lambda, [left, right]);
export const TTicket = (innerType: IType) => new ContainerType(TypeAtom.ticket, [innerType]);
export const TContract = (innerType: IType) => new ContainerType(TypeAtom.contract, [innerType]);
export const TSapling_state = (memo: number) => new ContainerType(TypeAtom.sapling_state, [memo]);
export const TSapling_transaction = (memo: number) => new ContainerType(TypeAtom.sapling_transaction, [memo]);
// Artificial Types
export const TRecord = (fields: Record<string, IType>, layout?: ILayout | Layout) =>
    new Type_Record(TypeAtom.record, fields, layout);
export const TVariant = (fields: Record<string, IType>, layout?: ILayout | Layout) =>
    new Type_Record(TypeAtom.variant, fields, layout);

const Types = {
    TUnknown,
    // Singleton types
    TUnit,
    TNat,
    TInt,
    TMutez,
    TString,
    TBool,
    TBytes,
    TAddress,
    TTimestamp,
    TChain_id,
    TBls12_381_fr,
    TBls12_381_g1,
    TBls12_381_g2,
    TKey,
    TKey_hash,
    TSignature,
    TOperation,
    TNever,
    // Container types
    TList,
    TSet,
    TOption,
    TPair,
    TMap,
    TBig_map,
    TLambda,
    TTicket,
    TContract,
    TSapling_state,
    TSapling_transaction,
    // Artificial Types
    TRecord,
    TVariant,
};

export default Types;
