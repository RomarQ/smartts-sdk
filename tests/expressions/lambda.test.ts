import {
    CallLambda,
    ContractStorage,
    Lambda,
    Nat,
    String,
    GreaterThan,
    GetVariable,
    FirstElement,
    SecondElement,
    Add,
    Int,
    Subtract,
    AsType,
} from '../../src/expression';
import { Contract, EntryPoint } from '../../src/core';
import { verifyContractCompilationOutput, verifyValueCompilationOutput } from '../util';
import { If, MatchVariant, NewVariable, Return, SetValue } from '../../src/statement';
import { TInt, TLambda, TNat, TPair, TString, TUnknown, TVariant } from '../../src/type';
import { LineInfo } from '../../src/misc/utils';

describe('Lambda expressions', () => {
    it('CallLambda', () => {
        const lambda = Lambda()
            .setInputType(TNat())
            .code((arg) => [
                If(GreaterThan(arg, Nat(2)))
                    .Then([Return(String('YES'))])
                    .Else([Return(String('NO'))]),
            ]);

        const contract = new Contract()
            .setStorageType(TString())
            .addEntrypoint(
                new EntryPoint('entry_point_1')
                    .setInputType(TNat())
                    .code((arg) => [SetValue(ContractStorage(), CallLambda(lambda, arg))]),
            );

        verifyContractCompilationOutput(contract);
    });
    it('Lambda with variant', () => {
        const lambda = Lambda()
            .setInputType(
                TVariant({
                    add: TPair(TInt(), TInt()),
                    subtract: TPair(TInt(), TInt()),
                }),
            )
            .code((arg) => [
                NewVariable('result', Int(0)),
                MatchVariant(arg)
                    .Case('add', (add_arg) => [
                        SetValue(GetVariable('result'), Add(FirstElement(add_arg), SecondElement(add_arg))),
                    ])
                    .Case('subtract', (subtract_arg) => [
                        SetValue(
                            GetVariable('result'),
                            Subtract(FirstElement(subtract_arg), SecondElement(subtract_arg)),
                        ),
                    ]),
                Return(GetVariable('result')),
            ]);

        verifyValueCompilationOutput(lambda);
    });
});
