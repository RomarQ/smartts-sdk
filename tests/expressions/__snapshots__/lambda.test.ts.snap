// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lambda expressions CallLambda 1`] = `
"
        (
            template_id (static_id 0 (\\"lambda.test.ts\\" 32))
            storage ()
            storage_type (\\"string\\")
            messages ((entry_point_1 True False False True (\\"lambda.test.ts\\" 35) ((set_type (params (\\"lambda.test.ts\\" 35)) \\"nat\\" (\\"lambda.test.ts\\" 40)) (set (data) (call_lambda (type_annotation (lambda 1 None False \\"lambda_arg\\" (\\"lambda.test.ts\\" 24) ((ifBlock (gt (lambdaParams 1 \\"lambda_arg\\" (\\"lambda.test.ts\\" 24) \\"nat\\") (literal (nat 2) (\\"lambda.test.ts\\" 26)) (\\"lambda.test.ts\\" 26)) ((result (literal (string \\"YES\\") (\\"lambda.test.ts\\" 26)) (\\"lambda.test.ts\\" 26))) (\\"lambda.test.ts\\" 26)) (elseBlock ((result (literal (string \\"NO\\") (\\"lambda.test.ts\\" 26)) (\\"lambda.test.ts\\" 26)))))) (lambda \\"nat\\" (unknown 0)) (\\"lambda.test.ts\\" 24)) (params (\\"lambda.test.ts\\" 35)) (\\"lambda.test.ts\\" 37)) (\\"lambda.test.ts\\" 37)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"lambda.test.ts\\" 32))
        )
        "
`;

exports[`Lambda expressions CallLambda 2`] = `
"{
    \\"micheline\\": \\"parameter (nat %entry_point_1);\\\\nstorage   string;\\\\ncode\\\\n  {\\\\n    CAR;\\\\n    PUSH nat 2;\\\\n    COMPARE;\\\\n    LT;\\\\n    IF\\\\n      {\\\\n        PUSH string \\\\\\"YES\\\\\\";\\\\n      }\\\\n      {\\\\n        PUSH string \\\\\\"NO\\\\\\";\\\\n      };\\\\n    NIL operation;\\\\n    PAIR;\\\\n  };\\",
    \\"json\\": [
        {
            \\"prim\\": \\"storage\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"string\\"
                }
            ]
        },
        {
            \\"prim\\": \\"parameter\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"nat\\",
                    \\"annots\\": [
                        \\"%entry_point_1\\"
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"code\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"CAR\\"
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"nat\\"
                            },
                            {
                                \\"int\\": \\"2\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"COMPARE\\"
                    },
                    {
                        \\"prim\\": \\"LT\\"
                    },
                    {
                        \\"prim\\": \\"IF\\",
                        \\"args\\": [
                            [
                                {
                                    \\"prim\\": \\"PUSH\\",
                                    \\"args\\": [
                                        {
                                            \\"prim\\": \\"string\\"
                                        },
                                        {
                                            \\"string\\": \\"YES\\"
                                        }
                                    ]
                                }
                            ],
                            [
                                {
                                    \\"prim\\": \\"PUSH\\",
                                    \\"args\\": [
                                        {
                                            \\"prim\\": \\"string\\"
                                        },
                                        {
                                            \\"string\\": \\"NO\\"
                                        }
                                    ]
                                }
                            ]
                        ]
                    },
                    {
                        \\"prim\\": \\"NIL\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"operation\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    }
                ]
            ]
        }
    ],
    \\"smartpy\\": \\"import smartpy as sp\\\\n\\\\nclass Contract(sp.Contract):\\\\n  def __init__(self):\\\\n    self.init_type(sp.TString)\\\\n\\\\n  @sp.entry_point\\\\n  def entry_point_1(self, params):\\\\n    sp.set_type(params, sp.TNat)\\\\n    def flambda_arg(lambda_arg):\\\\n      sp.if lambda_arg > 2:\\\\n        sp.result('YES')\\\\n      sp.else:\\\\n        sp.result('NO')\\\\n    self.data = sp.set_type_expr(sp.build_lambda(flambda_arg), sp.TLambda(sp.TNat, sp.TString))(params)\\\\n\\\\nsp.add_compilation_target(\\\\\\"visualtez_compilation\\\\\\", Contract())\\"
}"
`;

exports[`Lambda expressions Lambda with variant 1`] = `"(type_annotation (lambda 2 None False \\"lambda_arg\\" (\\"lambda.test.ts\\" 43) ((defineLocal \\"result\\" (literal (int 0) (\\"lambda.test.ts\\" 50)) True (\\"lambda.test.ts\\" 50)) (match_cases (lambdaParams 2 \\"lambda_arg\\" (\\"lambda.test.ts\\" 43) (variant ((add (tuple \\"int\\" \\"int\\")) (subtract (tuple \\"int\\" \\"int\\"))) (Some ((\\"add\\") (\\"subtract\\"))) (\\"\\" -1))) \\"__MATCH_1__\\" ((match (cases_arg \\"__MATCH_1__\\" (\\"lambda.test.ts\\" 50)) add \\"__MATCH_1___add\\" ((set (getLocal \\"result\\" (\\"lambda.test.ts\\" 50)) (add (first (variant_arg \\"__MATCH_1___add\\" (\\"lambda.test.ts\\" 50)) (\\"lambda.test.ts\\" 50)) (second (variant_arg \\"__MATCH_1___add\\" (\\"lambda.test.ts\\" 50)) (\\"lambda.test.ts\\" 50)) (\\"lambda.test.ts\\" 50)) (\\"lambda.test.ts\\" 50))) (\\"lambda.test.ts\\" 50)) (match (cases_arg \\"__MATCH_1__\\" (\\"lambda.test.ts\\" 50)) subtract \\"__MATCH_1___subtract\\" ((set (getLocal \\"result\\" (\\"lambda.test.ts\\" 50)) (sub (first (variant_arg \\"__MATCH_1___subtract\\" (\\"lambda.test.ts\\" 50)) (\\"lambda.test.ts\\" 50)) (second (variant_arg \\"__MATCH_1___subtract\\" (\\"lambda.test.ts\\" 50)) (\\"lambda.test.ts\\" 50)) (\\"lambda.test.ts\\" 50)) (\\"lambda.test.ts\\" 50))) (\\"lambda.test.ts\\" 50))) (\\"lambda.test.ts\\" 50)) (result (getLocal \\"result\\" (\\"lambda.test.ts\\" 50)) (\\"lambda.test.ts\\" 50)))) (lambda (variant ((add (tuple \\"int\\" \\"int\\")) (subtract (tuple \\"int\\" \\"int\\"))) (Some ((\\"add\\") (\\"subtract\\"))) (\\"\\" -1)) (unknown 0)) (\\"lambda.test.ts\\" 43))"`;

exports[`Lambda expressions Lambda with variant 2`] = `
"{
    \\"micheline\\": \\"{ IF_LEFT { UNPAIR; ADD } { UNPAIR; SUB } }\\",
    \\"json\\": [
        {
            \\"prim\\": \\"IF_LEFT\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"UNPAIR\\"
                    },
                    {
                        \\"prim\\": \\"ADD\\"
                    }
                ],
                [
                    {
                        \\"prim\\": \\"UNPAIR\\"
                    },
                    {
                        \\"prim\\": \\"SUB\\"
                    }
                ]
            ]
        }
    ]
}"
`;
